### ISSUE TITLE
[apex] AvoidSoqlInLoops false positive for SOQL with in For-Loop

### ISSUE THREAD

**Issue by ramachandra-mohan** [pmd/pmd#2598]
**Affects PMD Version:**  6.24.0

**Description:**

The rule [AvoidSoqlInLoops](https://pmd.github.io/latest/pmd_rules_apex_performance.html#avoidsoqlinloops) does report the following violation, when it shouldn't:

```java
public class Foo {
    public void bar() {
        for(NAMESPACE__CustomObject__c customObject : DatabaseUtility.query([Select Id FROM NAMESPACE__CustomObject__c where id IN :setIds])) // <-- violation is reported here
        {
            // …
        }
    }
}
```

This is a false-positive!

**Running PMD through:**  Command line

**Comment by adangel** [comment]
@ramachandra-mohan I've updated the issue description - Did I understand you correctly, that you reported a false-negative? That means: you expected, that PMD would report the violation for the above code, but didn't?

If that's the case, then this is not a bug, this is expected behavior - we even have a test case for this scenario: 

https://github.com/pmd/pmd/blob/eed1c6cab282287813fb8757fc9d9f9be68dd5b3/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidSoqlInLoops.xml#L92-L103

The reason is, that the SOQL query is not within the loop, it is executed only once. That's why this is not a problem.

@rsoesemann: Are my assumptions here correct? Do we need to change the rule?

**Comment by ramachandra-mohan** [comment]
When i execute the PMD from command line in linux system, PMD reports this as AvoidSoqlInLoops issue.

**Comment by rsoesemann** [comment]
Seems to a false positive as the reported code snippet is NOT an issue.

**Comment by adangel** [comment]
> When i execute the PMD from command line in linux system, PMD reports this as AvoidSoqlInLoops issue.

@ramachandra-mohan Thanks for clarifying - I totatlly misunderstood you - you were reporting a false-positive. I've updated the bug description accordingly.

Unfortunately, I can't reproduce the problem.
What exactly is the command line you are executing? And can you maybe share the whole file, where the violation is being reported?

**Comment by ramachandra-mohan** [comment]
@adangel I just identified the different between the above example and actual code. We are using utility method to verify list and return the list of sobjects. Could you please use below code to replicate the issue ? DatabaseUtils.query(List<SObject) returns List<SObject> after filtering the records.
```
for(NAMESPACE__CustomObject__c customObject : DatabaseUtility.query([Select Id FROM NAMESPACE__CustomObject__c where id IN :setIds])){
}
```

**Comment by adangel** [comment]
@ramachandra-mohan Thanks, now I can reproduce the issue.
